syntax = "proto3";

import "network/protos/ModelUpdate.proto";

service Initialize {
    rpc InitIdentity(Identity) returns (Ack) {};
    rpc InitDataset(Dataset) returns (Ack) {};
    rpc InitModel(Model) returns (Ack) {};
    rpc InitModelParameters(ModelParameters) returns (Ack) {};
    rpc InitLearningStrategy(LearningStrategy) returns (Ack) {};
    rpc RegisterNeighbors(NeighborSpec) returns (Ack) {};
    rpc StartLearning(Ack) returns (Ack) {};
};

message Identity {
    NetworkIdentity net_id = 1;
    int32 num_workers = 2;
    int32 seed = 3; // actor-specific seed
}

message Dataset {
    int32 dataset_id = 1; // enum id
    Partition partition = 2; // partitioning properties
};

message Partition {
    int32 partition_scheme_id = 1; // enum id
    int32 partition_index = 2; // index of partition in federated dataset
    int32 dataset_seed = 3; // seed used for data splitting and partitioning (same among actors)
    optional float partition_dirichlet_alpha = 4; // parameter for dirichlet partitioning
};

message Model {
    bytes model_config = 1; // serialized model configuration
    bytes optimizer_config = 2; // serialized optimizer configuration
};

message LearningStrategy {
    int32 learning_type_id = 1; // enum id of learning strategy
    SynchronizationStrategySpec sync_strat_spec = 2; // synchronization strategy specification
    CompressionStrategySpec compr_strat_spec = 3; // compression strategy specification
    PartialDeviceParticipationStrategySpec pdp_strat_spec = 4; // partial device participation strategy specification
};

message SynchronizationStrategySpec {
    int32 strategy_id = 1; // enum id
    optional float percentage = 2; // min percentage of updates from neighbors
    optional int32 amount = 3; // min amount of updates
    optional float timeout = 4; // timeout for waiting for model updates
};

message CompressionStrategySpec {
    int32 strategy_id = 1; // enum id of compression type
    optional int32 k = 2; // amount parameter for sparsification
    optional float percentage = 3; // percentage parameter for sparsification
    optional int32 precision = 4; // bit-precision parameter for quantization
};

message PartialDeviceParticipationStrategySpec {
    int32 strategy_id = 1; // enum id of partial device participation strategy
    optional int32 k = 2; // number of participants parameter
};

message NeighborSpec {
    repeated NetworkIdentity net_id = 1;
};
